### PR TITLE
refactor(core): disallow standalone components in importProvidersFrom calls

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This commit narrows down acceptable argument types of the
`importProvidersFrom` function. More specifically, it rejects
standalone components as a source of imports.

Based on the discussion from https://github.com/angular/angular/pull/45766/files#r858684934

### CODE REVIEW

File: packages/core/src/di/provider_collection.ts [hunk 1]
```diff
@@ -38,10 +40,25 @@ export type ImportProvidersSource =
...
+}
+
+export function internalImportProvidersFrom(
+    checkForStandaloneCmp: boolean, ...sources: ImportProvidersSource[]): Provider[] {
   const providersOut: SingleProvider[] = [];
   const dedup = new Set<Type<unknown>>();  // already seen types
   let injectorTypesWithProviders: InjectorTypeWithProviders<unknown>[]|undefined;
   deepForEach(sources, source => {
+    if (ngDevMode && checkForStandaloneCmp) {
+      const cmpDef = getComponentDef(source);
```
Comment (crisbeto) [inline packages/core/src/di/provider_collection.ts:53]: Drive-by comment: should this look for directive and pipe definitions as well?

Comment (pkozlowski-opensource) [reply packages/core/src/di/provider_collection.ts:53]: We could but those don't have imports and the operation will be noop - effectively making it equivalent to providing "random" type and the "existing machinery should take care of it".

**Comment by pkozlowski-opensource** [comment]
I really, really dislike the fact that extracting a function upsets the size-tracking checks... I will fix those after integrating review comments (if any).

File: packages/core/src/errors.ts [hunk 2]
```diff
@@ -67,6 +67,9 @@ export const enum RuntimeErrorCode {
   UNSAFE_VALUE_IN_SCRIPT = 905,
   MISSING_GENERATED_DEF = 906,
   TYPE_IS_NOT_STANDALONE = 907,
+
+  // importProvidersFrom Errors
+  IMPORT_PROVIDERS_FROM_STANDALONE = 1000,
```
Comment (JoostK) [inline packages/core/src/errors.ts:72]: Oooh, we might be getting duplicates with compiler errors codes now; the compiler starts at NG1000 and the runtime normally has three digits with a `0` prefix, so this would be `NG01000`.

File: packages/core/src/di/provider_collection.ts [hunk 3]
```diff
@@ -43,10 +45,25 @@ export type ImportProvidersSource =
...
+  return {ɵproviders: internalImportProvidersFrom(true, sources)};
+}
+
+export function internalImportProvidersFrom(
+    checkForStandaloneCmp: boolean, ...sources: ImportProvidersSource[]): Provider[] {
   const providersOut: SingleProvider[] = [];
   const dedup = new Set<Type<unknown>>();  // already seen types
   let injectorTypesWithProviders: InjectorTypeWithProviders<unknown>[]|undefined;
   deepForEach(sources, source => {
+    if (ngDevMode && checkForStandaloneCmp) {
```
[the remark this answers is not in the document]
Comment (pkozlowski-opensource) [reply packages/core/src/di/provider_collection.ts:57]: good call, changed.

File: packages/core/src/errors.ts (hunk 2 above)
Comment (pkozlowski-opensource) [reply packages/core/src/errors.ts]: ok, I see, changed.

**Review by alxhub — APPROVED** [review]
Reviewed-for: global-approvers

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 410d81f0e23e2c7f744ad27b2b2c0704171a2258.
